File: prognosia_backend/suggestions.py
import openai
import pdfplumber
import json
import os
import re
import csv
import sys
from datetime import datetime
from config import API_KEY
from PIL import Image

# Configura tu clave de API
client = openai.OpenAI(api_key=API_KEY)

# 🔹 Directorios y archivos importantes
ESTUDIOS_DIR = "./estudios"
PACIENTE_DIR = "./paciente"
DATA_JSON = os.path.join(PACIENTE_DIR, "data.json")
HISTORIA_ARCHIVO = os.path.join(PACIENTE_DIR, "historia")  # Puede ser .pdf o .txt
RESUMEN_ESTUDIOS = os.path.join(PACIENTE_DIR, "resumen_estudios.txt")
RESUMEN_HISTORIA = os.path.join(PACIENTE_DIR, "resumen_historia.txt")
SUGERENCIAS_JSON = os.path.join(PACIENTE_DIR, "sugerencias.json") 
METADATA_FILE = os.path.join(PACIENTE_DIR, "metadata.json")

# 🔹 Validar entrada de parámetros en ejecución
if len(sys.argv) < 3:
    print("⚠️ Uso: python suggestions.py \"medicamentos\" \"sintomas\"")
    sys.exit(1)

# Función para detectar imágenes en el directorio de estudios
def detectar_imagenes():
    return [f for f in os.listdir(ESTUDIOS_DIR) if f.lower().endswith((".png", ".jpg", ".jpeg"))]

def verificar_directorio(directorio):
    if not os.path.exists(directorio):
        os.makedirs(directorio)

def obtener_metadata():
    if os.path.exists(METADATA_FILE):
        with open(METADATA_FILE, "r", encoding="utf-8") as file:
            return json.load(file)
    return {}

def guardar_metadata(metadata):
    with open(METADATA_FILE, "w", encoding="utf-8") as file:
        json.dump(metadata, file, indent=4)

def obtener_fecha_modificacion(archivo):
    return os.path.getmtime(archivo) if os.path.exists(archivo) else 0

def archivos_cambiaron(directorio, ultima_actualizacion):
    for archivo in os.listdir(directorio):
        ruta_archivo = os.path.join(directorio, archivo)
        if obtener_fecha_modificacion(ruta_archivo) > ultima_actualizacion:
            return True
    return False

# 🔹 Función para extraer texto de archivos PDF
def extraer_texto_archivo(ruta_archivo):
    if ruta_archivo.endswith(".pdf"):
        with pdfplumber.open(ruta_archivo) as pdf:
            return "\n".join([pagina.extract_text() for pagina in pdf.pages if pagina.extract_text()])
    elif ruta_archivo.endswith(".txt"):
        with open(ruta_archivo, "r", encoding="utf-8") as file:
            return file.read().strip()
    elif ruta_archivo.endswith(".csv"):
        with open(ruta_archivo, "r", encoding="utf-8") as file:
            return "\n".join([", ".join(row) for row in csv.reader(file)])
    return ""

# 🔹 Función para actualizar el resumen de estudios (TXT, CSV o PDF)
def actualizar_resumen_estudios(metadata):
    if archivos_cambiaron(ESTUDIOS_DIR, metadata.get("ultima_actualizacion_estudios", 0)):
        texto_total = "".join(extraer_texto_archivo(os.path.join(ESTUDIOS_DIR, archivo)) for archivo in os.listdir(ESTUDIOS_DIR))
        with open(RESUMEN_ESTUDIOS, "w", encoding="utf-8") as file:
            file.write(texto_total[:8000])
        metadata["ultima_actualizacion_estudios"] = datetime.now().timestamp()
        guardar_metadata(metadata)

def actualizar_resumen_historia(metadata):
    historia_txt = HISTORIA_ARCHIVO + ".txt"
    historia_pdf = HISTORIA_ARCHIVO + ".pdf"
    archivos = [historia_pdf, historia_txt]
    
    if any(obtener_fecha_modificacion(a) > metadata.get("ultima_actualizacion_historia", 0) for a in archivos):
        texto = extraer_texto_archivo(historia_pdf) if os.path.exists(historia_pdf) else extraer_texto_archivo(historia_txt)
        with open(RESUMEN_HISTORIA, "w", encoding="utf-8") as file:
            file.write(texto[:8000])
        metadata["ultima_actualizacion_historia"] = datetime.now().timestamp()
        guardar_metadata(metadata)

# 🔹 Función para cargar y validar los datos del paciente desde JSON
def cargar_datos_paciente():
    verificar_directorio(PACIENTE_DIR)

    if not os.path.exists(DATA_JSON):
        print("⚠️ No se encontró 'data.json'. Creando uno nuevo...")
        with open(DATA_JSON, "w", encoding="utf-8") as file:
            json.dump({"nombre": "No definido"}, file, indent=4)
        return None

    try:
        with open(DATA_JSON, "r", encoding="utf-8") as file:
            return json.load(file)
    except Exception as e:
        print(f"⚠️ Error al leer el archivo JSON: {e}")
        return None

def generar_resumen_ia(resumen):
    imagenes = detectar_imagenes()
    model = "gpt-4o" if imagenes else "gpt-4o-mini"

    with open(RESUMEN_ESTUDIOS, "r", encoding="utf-8") as file:
        contenido = file.read()
    
    messages = [{"role": "user", "content": f"Genera un resumen claro y breve de este contenido: {contenido}"}]
    
    if imagenes:
        for img in imagenes:
            with open(os.path.join(ESTUDIOS_DIR, img), "rb") as image_file:
                messages.append({"role": "user", "content": {"type": "image", "image": image_file.read()}})
    
    response = client.chat.completions.create(model=model, messages=messages)
    return response.choices[0].message.content

def guardar_sugerencia(sugerencia, medicamentos, sintomas):
    data = {
        "fecha": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "medicamentos": medicamentos,
        "sintomas": sintomas,
        "sugerencia": sugerencia
    }
    with open(SUGERENCIAS_JSON, "w", encoding="utf-8") as file:
        json.dump(data, file, indent=4, ensure_ascii=False)

# 🔹 Función para solicitar estudios médicos sugeridos
def sugerir_estudios(datos_medicos, medicamentos, sintomas, resumen_estudios, resumen_historia):
    prompt = f"""
    Basado en la siguiente información resumida del paciente, sugiere una lista de estudios médicos recomendados.

    📌 **Datos médicos generales**
    - Nombre: {datos_medicos.get("nombre", "No especificado")}
    - Edad: {datos_medicos.get("edad", "No especificado")} años
    - Sexo: {datos_medicos.get("sexo", "No especificado")}
    - Enfermedades crónicas: {datos_medicos.get("enfermedades_cronicas", "No especificado")}

    💊 **Medicamentos actuales**
    {medicamentos}

    🔬 **Síntomas actuales reportados**
    {sintomas}

    📄 **Resumen de estudios médicos previos**
    {resumen_estudios}

    📖 **Resumen de historia clínica**
    {resumen_historia}

    **Responde con una lista de estudios sugeridos, su breve descripción y por qué son importantes. 
    Estos deben ser faciles de entender por el usuario y no debe ser muy especifico. No debe asustar o hacer preocupar al usuario.
    Ten en cuenta para esto ademas de los datos clinicos, personales, sintomas y medicamentos tambien la frecuencia con la que las
    personas de ese grupo deben realizarse estudios de distintos tipos y la ultima vez que se los haya realizado.
    Tambien debes darle importancia a los sintomas actuales y los datos previos.**
    """
    #**Responde con una lista de estudios sugeridos, sin explicaciones adicionales.**
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": prompt}]
    )
    return response.choices[0].message.content

# Ejecutar el programa
def suggest(medicamentos, sintomas):

    if not medicamentos:
        medicamentos = ""
    if not sintomas:
        sintomas = ""

    metadata = obtener_metadata()
    datos_paciente = cargar_datos_paciente()

    actualizar_resumen_estudios(metadata)
    actualizar_resumen_historia(metadata)

    with open(RESUMEN_ESTUDIOS, "r", encoding="utf-8") as file:
        resumen_estudios = generar_resumen_ia(file.read())

    with open(RESUMEN_HISTORIA, "r", encoding="utf-8") as file:
        resumen_historia = generar_resumen_ia(file.read())

    resultado = sugerir_estudios(datos_paciente, medicamentos, sintomas, resumen_estudios, resumen_historia)
    guardar_sugerencia(resultado, medicamentos, sintomas)

    print("\n📋 **Estudios sugeridos:**\n")
    print(resultado)

File: prognosia-frontend/services/ai_services/suggestions.js
 function suggest(medicamentos, sintomas){
    /* 
    * suggestions.js
    * 
    * Ejemplo de traducción de suggestions.py (Python) a JavaScript (Node.js).
    * 
    * Instalaciones necesarias (vía npm):
    *  npm install openai pdf-parse csv-parse fs path
    *
    * Ajusta las rutas, modelos y lógicas según sea necesario.
    */

    const fs = require("fs");
    const path = require("path");
    const { OpenAI } = require("openai");
    const pdfParse = require("pdf-parse");
    const { parse } = require("csv-parse/sync");
    const config = require("./config.json");1
    // 🔹 Configuración de OpenAI
    const API_KEY = config.API_KEY;

    // Instancia de OpenAI
    const client = new OpenAI({ apiKey: API_KEY})


    // 🔹 Directorios y archivos importantes
    const ESTUDIOS_DIR = "./estudios";
    const PACIENTE_DIR = "./paciente";
    const DATA_JSON = path.join(PACIENTE_DIR, "data.json");
    const HISTORIA_ARCHIVO = path.join(PACIENTE_DIR, "historia");  // Puede ser .pdf o .txt
    const RESUMEN_ESTUDIOS = path.join(PACIENTE_DIR, "resumen_estudios.txt");
    const RESUMEN_HISTORIA = path.join(PACIENTE_DIR, "resumen_historia.txt");
    const SUGERENCIAS_JSON = path.join(PACIENTE_DIR, "sugerencias.json");
    const METADATA_FILE = path.join(PACIENTE_DIR, "metadata.json");

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 1. OBTENCIÓN DE PARÁMETROS DESDE LA CLI ───────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    // if (process.argv.length < 4) {
    // console.log("⚠️ Uso: node suggestions.js \"<medicamentos>\" \"<sintomas>\"");
    // process.exit(1);
    // }
    const medicamentosArg = medicamentos || "";
    const sintomasArg = sintomas || "";

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 2. FUNCIONES AUXILIARES ──────────────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────

    // Verificar/crear directorio si no existe
    function verificarDirectorio(directorio) {
    if (!fs.existsSync(directorio)) {
        fs.mkdirSync(directorio, { recursive: true });
    }
    }

    // Cargar metadata desde METADATA_FILE
    function obtenerMetadata() {
    if (fs.existsSync(METADATA_FILE)) {
        const contenido = fs.readFileSync(METADATA_FILE, "utf-8");
        return JSON.parse(contenido);
    }
    return {};
    }

    // Guardar metadata en METADATA_FILE
    function guardarMetadata(metadata) {
    fs.writeFileSync(METADATA_FILE, JSON.stringify(metadata, null, 2), "utf-8");
    }

    // Obtener fecha de modificación de un archivo (en milisegundos)
    function obtenerFechaModificacion(archivo) {
    try {
        return fs.statSync(archivo).mtimeMs;
    } catch (error) {
        return 0;
    }
    }

    // Verificar si algún archivo en un directorio cambió después de cierta marca de tiempo
    function archivosCambiaron(directorio, ultimaActualizacion) {
    if (!fs.existsSync(directorio)) return false;
    const archivos = fs.readdirSync(directorio);
    return archivos.some(archivo => {
        const rutaArchivo = path.join(directorio, archivo);
        return obtenerFechaModificacion(rutaArchivo) > ultimaActualizacion;
    });
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 3. EXTRACCIÓN DE TEXTO DE ARCHIVOS ────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function extraerTextoArchivo(rutaArchivo) {
    if (!fs.existsSync(rutaArchivo)) return "";

    if (rutaArchivo.endsWith(".pdf")) {
        // PDF
        const data = fs.readFileSync(rutaArchivo);
        try {
        const pdfData = await pdfParse(data);
        return pdfData.text;
        } catch (error) {
        console.error("Error al procesar PDF:", error);
        return "";
        }
    } else if (rutaArchivo.endsWith(".txt")) {
        // TXT
        return fs.readFileSync(rutaArchivo, "utf-8").trim();
    } else if (rutaArchivo.endsWith(".csv")) {
        // CSV
        const contenido = fs.readFileSync(rutaArchivo, "utf-8");
        const registros = parse(contenido);
        // Unir filas con comas
        return registros.map(row => row.join(", ")).join("\n");
    }

    return "";
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 4. ACTUALIZAR RESUMEN DE ESTUDIOS ─────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function actualizarResumenEstudios(metadata) {
    const ultimaActualizacion = metadata.ultima_actualizacion_estudios || 0;

    if (archivosCambiaron(ESTUDIOS_DIR, ultimaActualizacion)) {
        // Extraer texto de todos los archivos en ESTUDIOS_DIR y concatenarlos
        const archivos = fs.readdirSync(ESTUDIOS_DIR);
        let textoTotal = "";

        for (const archivo of archivos) {
        const rutaArchivo = path.join(ESTUDIOS_DIR, archivo);
        const texto = await extraerTextoArchivo(rutaArchivo);
        textoTotal += texto;
        }

        // Limitar a 8000 caracteres
        const textoLimitado = textoTotal.slice(0, 8000);
        fs.writeFileSync(RESUMEN_ESTUDIOS, textoLimitado, "utf-8");

        metadata.ultima_actualizacion_estudios = Date.now();
        guardarMetadata(metadata);
    }
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 5. ACTUALIZAR RESUMEN DE HISTORIA ─────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function actualizarResumenHistoria(metadata) {
    const historiaTxt = HISTORIA_ARCHIVO + ".txt";
    const historiaPdf = HISTORIA_ARCHIVO + ".pdf";
    const ultimaActualizacion = metadata.ultima_actualizacion_historia || 0;

    const modTxt = obtenerFechaModificacion(historiaTxt);
    const modPdf = obtenerFechaModificacion(historiaPdf);

    // Si cualquiera de estos archivos es más reciente que la última actualización
    if (modTxt > ultimaActualizacion || modPdf > ultimaActualizacion) {
        let texto = "";
        if (fs.existsSync(historiaPdf)) {
        texto = await extraerTextoArchivo(historiaPdf);
        } else if (fs.existsSync(historiaTxt)) {
        texto = await extraerTextoArchivo(historiaTxt);
        }
        // Limitar a 8000 chars
        const textoLimitado = texto.slice(0, 8000);
        fs.writeFileSync(RESUMEN_HISTORIA, textoLimitado, "utf-8");

        metadata.ultima_actualizacion_historia = Date.now();
        guardarMetadata(metadata);
    }
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 6. CARGAR DATOS PACIENTE DESDE JSON ───────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    function cargarDatosPaciente() {
    verificarDirectorio(PACIENTE_DIR);

    if (!fs.existsSync(DATA_JSON)) {
        console.log("⚠️ No se encontró 'data.json'. Creando uno nuevo...");
        const dataInicial = { nombre: "No definido" };
        fs.writeFileSync(DATA_JSON, JSON.stringify(dataInicial, null, 2), "utf-8");
        return dataInicial;
    }

    try {
        const contenido = fs.readFileSync(DATA_JSON, "utf-8");
        return JSON.parse(contenido);
    } catch (error) {
        console.log("⚠️ Error al leer el archivo JSON:", error);
        return null;
    }
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 7. GENERAR RESUMEN VIA OPENAI ─────────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function generarResumenIA(contenido) {
        const model = "gpt-4o-mini"; // o "gpt-4", etc.

        const messages = `Genera un resumen claro y breve de este contenido:\n\n${contenido}`;

        // Verificar si hay una imagen en el directorio del paciente
        const imagenes = fs.readdirSync(PACIENTE_DIR).filter(file => {
            const ext = path.extname(file).toLowerCase();
            return ext === ".jpg" || ext === ".jpeg" || ext === ".png";
        });

        if (imagenes.length > 0) {
            const rutaImagen = path.join(PACIENTE_DIR, imagenes[0]); // Tomar la primera imagen encontrada
            try {
                const imagenBuffer = fs.readFileSync(rutaImagen);
                const imagenBase64 = imagenBuffer.toString("base64");
                messages.push( `Además, considera esta imagen codificada en base64:\n\n${imagenBase64}`);
            } catch (error) {
                console.error("Error al codificar la imagen en base64:", error);
            }
        }

        try {
            const response = await client.responses.create({
                model: model,
                input: messages
            });
            // Retorna el texto de la respuesta
            return response.output_text;
        } catch (error) {
            console.error("Error en generarResumenIA:", error);
            return "";
        }
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 8. GUARDAR SUGERENCIA EN JSON ─────────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    function guardarSugerencia(sugerencia, medicamentos, sintomas) {
    const data = {
        fecha: new Date().toISOString().replace("T", " ").split(".")[0],
        medicamentos: medicamentos,
        sintomas: sintomas,
        sugerencia: sugerencia
    };

    fs.writeFileSync(SUGERENCIAS_JSON, JSON.stringify(data, null, 2, "utf-8"), "utf-8");
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 9. SUGERIR ESTUDIOS VIA OPENAI ────────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function sugerirEstudios(
    datosMedicos, 
    medicamentos, 
    sintomas, 
    resumenEstudios, 
    resumenHistoria
    ) {
    // Aquí podrías usar GPT-4 si quieres
    const model = "gpt-4o-mini"; // Ajusta según tus necesidades

    const prompt = `
    Basado en la siguiente información resumida del paciente, sugiere una lista de estudios médicos recomendados.

    📌 **Datos médicos generales**
    - Nombre: ${datosMedicos?.nombre ?? "No especificado"}
    - Edad: ${datosMedicos?.edad ?? "No especificado"} años
    - Sexo: ${datosMedicos?.sexo ?? "No especificado"}
    - Enfermedades crónicas: ${datosMedicos?.enfermedades_cronicas ?? "No especificado"}

    💊 **Medicamentos actuales**
    ${medicamentos}

    🔬 **Síntomas actuales reportados**
    ${sintomas}

    📄 **Resumen de estudios médicos previos**
    ${resumenEstudios}

    📖 **Resumen de historia clínica**
    ${resumenHistoria}

    Responde con una lista de estudios sugeridos, su breve descripción y por qué son importantes. 
    Ten en cuenta tanto los datos clínicos, personales, síntomas y medicamentos, como la frecuencia recomendada para el grupo del paciente.
    Evita alarmar al paciente e intenta mantener el lenguaje sencillo. Evita cualquier saludo o despedida.
    `;

    try {
        const response = await client.responses.create({
        model: model,
        input: prompt
        });
        return response.output_text;
    } catch (error) {
        console.error("Error en sugerirEstudios:", error);
        return "";
    }
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 10. FUNCIÓN PRINCIPAL ─────────────────────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    async function main(medicamentos, sintomas) {
    const metadata = obtenerMetadata();
    const datosPaciente = cargarDatosPaciente();

    // Actualiza y genera los resúmenes en disco
    await actualizarResumenEstudios(metadata);
    await actualizarResumenHistoria(metadata);

    // Carga el contenido actual de los resúmenes
    let resumenEstudiosContenido = "";
    let resumenHistoriaContenido = "";

    if (fs.existsSync(RESUMEN_ESTUDIOS)) {
        resumenEstudiosContenido = fs.readFileSync(RESUMEN_ESTUDIOS, "utf-8");
    }
    if (fs.existsSync(RESUMEN_HISTORIA)) {
        resumenHistoriaContenido = fs.readFileSync(RESUMEN_HISTORIA, "utf-8");
    }

    // Genera un nuevo resumen con la IA para cada uno
    const resumenEstudios = await generarResumenIA(resumenEstudiosContenido);
    const resumenHistoria = await generarResumenIA(resumenHistoriaContenido);

    // Solicita sugerencias de estudios con toda la data
    const resultado = await sugerirEstudios(
        datosPaciente,
        medicamentos,
        sintomas,
        resumenEstudios,
        resumenHistoria
    );

    // Guarda la sugerencia en un JSON
    guardarSugerencia(resultado, medicamentos, sintomas);

    console.log("\n📋 **Estudios sugeridos:**\n");
    console.log(resultado);
    }

    // ─────────────────────────────────────────────────────────────────────────────
    // ──────────────── 11. EJECUCIÓN DE LA FUNCIÓN PRINCIPAL ─────────────────────
    // ─────────────────────────────────────────────────────────────────────────────
    main(medicamentosArg, sintomasArg)
    .then(() => {
        // Éxito
    })
    .catch((err) => {
        console.error("Ocurrió un error:", err);
    });
}

module.exports = suggest;